game-core/src/main/java/games/strategy/triplea/ui/BattleDisplay.java

```diff
@@ -340,43 +349,66 @@ void notifyRetreat(final Collection<Unit> retreating) {
     attackerModel.notifyRetreat(retreating);
   }
 
+  @Nullable
   Territory getRetreat(
       final String message, final Collection<Territory> possible, final boolean submerge) {
-    return (!submerge || possible.size() > 1)
-        ? getRetreatInternal(message, possible)
-        : getSubmerge(message);
-  }
-
-  private Territory getSubmerge(final String message) {
     if (SwingUtilities.isEventDispatchThread()) {
       throw new IllegalStateException("Should not be called from dispatch thread");
     }
-    final AtomicReference<Territory> retreatTo = new AtomicReference<>();
-    final CountDownLatch latch = new CountDownLatch(1);
-    final Action action =
-        SwingAction.of(
-            "Submerge Subs?",
-            e -> {
-              actionButton.setEnabled(false);
-              if (showSubmergeDialog(message, retreatTo)) {
-                actionButton.setAction(nullAction);
-                latch.countDown();
-              }
-              actionButton.setEnabled(true);
-            });
+    final CompletableFuture<Territory> future = new CompletableFuture<>();
+    final String title;
+    final Supplier<RetreatResult> supplier;
+    if (!submerge || possible.size() > 1) {
+      title = "Retreat?";
+      supplier = () -> showRetreatDialog(message, possible);
+    } else {
+      title = "Submerge Subs?";
+      supplier = () -> showSubmergeDialog(message);
+    }
+    final Action action = getPlayerAction(title, supplier, future);
```

Comment:
I think you missed the other part of my comment where I suggested not having this getPlayerAction() method and just inlining it here with the changes above, since it's only called once now.

Reply:
True, but if the method gets much longer code climate will complain.
That's why the method exists.